Suggest imports when using an undefined variable

Corecii opened the pull request:
Implements #122 

This PR adds suggested imports as quick actions when you use an undefined variable.

<details><summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/1669436/144704310-e01eb643-a465-4edc-9434-fddeb084fd75.png) ![image](https://user-images.githubusercontent.com/1669436/144704339-4072df96-f82c-4443-8590-da8e32ddae35.png)

![image](https://user-images.githubusercontent.com/1669436/144704376-4d3a87a2-603c-483b-9bd0-50867bcc84f5.png)

![image](https://user-images.githubusercontent.com/1669436/144704403-13747866-5258-4359-a2d2-501d8471ab80.png)


</details>

It accomplishes this by searching the rojo tree for matching modules then getting relative and absolute paths to the target modules. ~~This feature will prefer to show relative paths over absolute paths, but will use absolute paths when going through a service or through `game` is necessary.~~

This PR adds ~~two~~ three settings:
* `importScriptChildren`: Controls whether suggested require paths are allowed to index into scripts. It's common to treat individual scripts as their own resource and all of their children as "private" to that script. When following that convention, you will usually re-export anything you want to be publicly accessible.
  With `importScriptChildren = false`, Roblox LSP will still suggest relative require paths underneath a script that don't index into any new child scripts (as seen in screenshot 1 above).
* `importIgnore`: A set of globs that should never be suggested for automatic import. This is good when using something like Wally which creates many "package links" that the LSP would otherwise see as duplicate module names you could import. This is also shown above in screenshot 2.
* `importPathType`: A choice of which require paths you want the LSP to suggest to you, with choices ranging from only relative paths to only absolute paths and in-between.

Sorry that this is rather large! I found the codebase a bit hard to navigate, so I'm not sure if I've put everything in the best place or if an existing implementations exist for anything I've done. I appreciate any feedback necessary to clean this up and get it merged in!

Corecii commented:
This is currently just part of the `undefined-global` diagnostic. I have two thoughts about this:
1. It would be nice to be able to disable this "suggested imports" feature alone without disabling the diagnostic
2. I don't actually like using the built-in diagnostics. I prefer to use selene, but I still want to use this "suggested imports" feature.

What do you think is the best route to achieving these goals? (You're more familiar with the codebase and the direction of the project than I am!)

NightrainsRbx commented:
I love it! It remembers me when I was coding in C# and I never imported anything manually.

> It would be nice to be able to disable this "suggested imports" feature alone without disabling the diagnostic

I don't think this is necessary since you can decide whether to use it or not.

> I don't actually like using the built-in diagnostics. I prefer to use selene, so I usually disable all diagnostics. I still want to use this "suggested imports" feature with diagnostics disabled.

The way this works as a quick fix is perfect since that's how it works in most languages, so it can stay this way, however, there can be an alternative way to do it as a refactor. When implementing it as a refactor, you would have to check if the variable selected is an undefined global, the rest is almost the same.

I have some ideas:

1. Instead of deciding whether to use relative or absolute paths, you can use both and let the user choose, that is, two quick fixes/refactors.
2. To be consistent with how services are auto-imported, you should find in the AST a require and then place the import below it.
3. This could be a quick fix of its own diagnostic instead of `undefined-global`, this diagnostic will show up on undefined globals that can be auto-imported, this way people will notice it without opening Quick Fix. The diagnostic wouldn't be a warning but a hint or information.

NightrainsRbx commented on server/script/core/code-action.lua:
```diff
@@ -137,7 +142,222 @@ local function changeVersion(uri, version, results)
+end
+
+local function solveUndefinedGlobalImport(uri, diag, results)
+    local ast    = files.getAst(uri)
+    local offset = files.offsetOfWord(uri, diag.range.start)
+    local name = guide.eachSourceContain(ast.ast, offset, function (source)
+        if source.type ~= 'getglobal' then
+            return
+        end
+
+        return guide.getKeyName(source)
+    end)
```
This is duplicate code from `solveUndefinedGlobal`

Corecii replied:
This is now its own diagnostic, so I've left this duplicated. There's similar duplication between `solveUndefinedGlobal` and `solveLowercaseGlobal`. It's a small enough bit of code that I don't think it benefits much from being de-duplicated.

If we do de-duplicate it though, we should do it across this whole file -- maybe a function like `getKeyNameAtOffset(ast, offset, sourceType)`?

NightrainsRbx replied:
That's ok then.

NightrainsRbx commented on server/script/core/code-action.lua:
```diff
@@ -137,7 +142,222 @@ local function changeVersion(uri, version, results)
+end
+
 local function solveUndefinedGlobal(uri, diag, results)
+    solveUndefinedGlobalImport(uri, diag, results)
```
Move this above [`markGlobal(uri, name, results)`](https://github.com/Corecii/RobloxLsp/blob/e16862e97b09ce933cf5aed3673eff47138c746c/server/script/core/code-action.lua#L368) and pass `name` as an argument.

Corecii replied:
I've made this its own diagnostic now, so this suggestion is no longer relevant 👍

NightrainsRbx commented:
I think `importScriptChildren` and `importIgnore` should be in to `robloxLsp.misc` instead of `robloxLsp.diagnostics`.

Corecii commented:
Planning to address feedback here over this coming week. Thank you!

> Instead of deciding whether to use relative or absolute paths, you can use both and let the user choose, that is, with multiple quick fixes/refactors.

I originally did this, but I didn't like it because it doubled the available options, and in my case I wouldn't use half of them. Would you be okay if there as an additional setting for this?

> Very optional and probably unnecessary, but this could be a quick fix of its own diagnostic instead of `undefined-global`, this diagnostic will show up on undefined globals that can be auto-imported, this way people will notice it without opening Quick Fix. The diagnostic wouldn't be a warning but a hint or information.

This sounds good, and I almost did this originally. I was worried that showing both `undefined-global` and some other diagnostic might be weird and unnecessary. This is the best way to get the functionality I'm hoping for with the least work right now though, so I think I'll do it.

Corecii commented:
Sorry I've been slow on this! I was unexpectedly busy on the week I planned to finish this up, then I went on holiday break and barely even looked at my computer for two weeks. I'm back now and planning to finish this up soon!

Corecii commented:
I've addressed your feedback and I've done some cleanup by moving import-related functions to their own file.

This should be ready for your next review and potentially merging in!

---

> I think `importScriptChildren` and `importIgnore` should be in to `robloxLsp.misc` instead of `robloxLsp.diagnostics`.

Done!

> Very optional and probably unnecessary, but this could be a quick fix of its own diagnostic instead of `undefined-global`, this diagnostic will show up on undefined globals that can be auto-imported, this way people will notice it without opening Quick Fix. The diagnostic wouldn't be a warning but a hint or information.

Done!

> To be consistent with how services are auto-imported, you should find in the AST a require and then place the import below it.

Done!

> Instead of deciding whether to use relative or absolute paths, you can use both and let the user choose, that is, with multiple quick fixes/refactors.

I added a setting for this. I personally don't want to see absolute paths as I don't like the "quick fix" menu being cluttered. Users can choose a setting here that fits their conventions, including one to show both relative and absolute paths.

NightrainsRbx commented:
Should it use `robloxLsp.workspace.ignoreDir` instead of having its own setting?

Corecii commented:
I'm not sure. I can think of good reasons for this to go either way. If this is something you're certain you'd like me to change then I'm happy to change it.

Corecii commented:
I've added a default for the `importIgnore` setting, as I expect it to be common to ignore the wally index. If you don't ignore it then roblox lsp returns a ton of results that aren't useful.

If you decide it's better to use the `workspace.ignoreDir` setting:
1. Just let me know! I have a commit ready.
2. Let me know if you want me to add the `**/__index/**` default to `ignoreDir` or not.

NightrainsRbx commented on package.nls.json:
```diff
@@ -46,6 +46,9 @@
     "config.misc.goToScriptLink": "If enabled, shows a link to the source script when requiring a module.",
+    "config.misc.importScriptChildren": "Suggest require paths that index into scripts.",
+    "config.misc.importIgnore": "Globs to never suggest when suggesting require paths. For example: `**/_Index/**` to ignore wally's index folder. Case sensitive. Acts on the absolute path of files.",
+    "config.misc.importPathType": "Whether to show relative or absolute require paths.",
```
Add descriptions to enums. For example: 
```json
"config.misc.importPathType.Only Relative Paths": "Only import paths relative to the current script."
```

Corecii replied:
Ah, I didn't realize this was possible! Will do.

NightrainsRbx commented:
I think it should have its own setting, since most people when using `robloxLsp.workspace.ignoreDir` ignores the entire packages\vendor folder. Having `/__Index/` ignored by default is a good idea and could be by default in `robloxLsp.workspace.ignoreDir` as well.

I've been testing it for a while and it works fine, but when I import libraries like Roact from ReplicatedStorage, or other modules, I don't like it using the relative path, maybe it should prefer Absolute Paths by default, or there should be a different way to decide where to use Absolute or Relative paths, for example, maybe it could use the shortest one?

Corecii commented:
> I've been testing it for a while and it works fine, but when I import libraries like Roact from ReplicatedStorage, or other modules, I don't like it using the relative path, maybe it should prefer Absolute Paths by default

We can change the default setting here.

An `Absolute First` option doesn't make sense because there should *always* be an absolute path, so with `Absolute First` the relative paths would never show (effectively `Absolute Only`).

I will be personally making use of `Relative First`. We all have codebases structured differently, and have different conventions. I'll be using `Relative First` for a game, but it's also a good setting for developing packages or plugins because both are typically contained in a single folder and should not be requiring from the root game tree.

> or there should be a different way to decide where to use Absolute or Relative paths, for example, maybe it could use the shortest one?

I can add an option `Shortest First` to the `importPathType` enum for this.

It sounds like for you either `Shortest First` or `Both` are best for your codebase and conventions.

---

One idea I had is *another* glob that limits what can be required relatively. You could set this to e.g. `Packages/**` and now Roblox LSP will always use absolute paths for files in that folder.

I think we should:
1. Implement `Shortest First` or similar
2. Potentially implement "always absolute" glob settings
3. Decide on a good default setting for `importPathType`
4. Release this as a first version. Codifying `require` conventions can be *hard*, but we have to start somewhere!

NightrainsRbx commented:
Ok then this can be merged!

Corecii commented:
There were a few changes (discussed above) that I wanted to get in before this was merged. I've set up a small pr here with those changes: https://github.com/NightrainsRbx/RobloxLsp/pull/134
